src/main/java/com/zendesk/maxwell/producer/MaxwellKafkaProducer.java

```diff
@@ -75,9 +86,25 @@ public void onCompletion(RecordMetadata md, Exception e) {
 				LOGGER.debug("   " + position);
 				LOGGER.debug("");
 			}
+			cc.markCompleted();
+		}
+	}
+
+	private void publishFallback(RecordMetadata md, Exception e) {
+		// When publishing a fallback record, make a callback
+		// with no fallback topic to avoid infinite loops
+		KafkaCallback cb = new KafkaCallback(cc, position, key, json,
+			succeededMessageCount, failedMessageCount, succeededMessageMeter,
+			failedMessageMeter, null, context, producer);
+		try {
+			producer.sendFallbackAsync(fallbackTopic, key, cb, e);
```

Comment:
i think maybe the logic is too fractured between this call point and `sendFallbackAsync` -- it seems like they're splitting the duties a bit too much.  Maybe here we call `producer.fallback(this, e)` and then it does the rest of the work of creating a new callback etc?  

sorry to go another rev, I'm just having trouble following all the tendrils of logic here.

Reply:
So it's easy enough to move the try/catch out of here, you just need to pass the `RecordMetadata` on to the publisher (done in 9cb2a91ceba0992e058c3ed72e7e36382053cb9e). But if you want to move the creation of the callback into the publisher, you'd need to make the ~9 private fields used for construction visible to the publisher (or wrap them with a content object 😛 ).

You'd also need to move the assertions in https://github.com/zendesk/maxwell/blob/3ecbebd0e48b8e8706f7e2a7d71e759117f33d72/src/test/java/com/zendesk/maxwell/producer/KafkaCallbackTest.java#L116-L118 (which check that the fallback topic is `"dead_letters"` on the first error, then `null` on the second) into a new test in MaxwellKafkaProducerWorkerTest. Which is not hard, just tedious due to all the mocking.

So I can do that, but I'm not convinced it would end up particularly nice. To be honest, I really don't mind how it's structured so long as it works and is tested. Would you prefer to have a go at refactoring this branch however you'd like it to be structured?